File: demos/kotlin/algorithms_demo.kt
data class Score(val name: String, val value: Int)

fun bfs(graph: Map<String, List<String>>, start: String): List<String> {
    val order = mutableListOf<String>()
    val seen = mutableSetOf<String>()
    val queue = ArrayDeque(listOf(start))
    while (queue.isNotEmpty()) {
        val node = queue.removeFirst()
        if (!seen.add(node)) continue
        order += node
        queue.addAll(graph[node].orEmpty())
    }
    return order
}

fun main() {
    val ranked = listOf(Score("Lin", 91), Score("Ada", 95), Score("Kai", 91))
        .sortedWith(compareByDescending<Score> { it.value }.thenBy { it.name })
    val numbers = listOf(2, 5, 8, 13, 21)
    val graph = mapOf("A" to listOf("B", "C"), "B" to listOf("D"), "C" to listOf("D"))

    println("sorted=${ranked.joinToString(",") { "${it.name}:${it.value}" }}")
    println("binary-search-13=${numbers.binarySearch(13)}")
    println("bfs=${bfs(graph, "A").joinToString(",")}")
}

File: demos/ruby/algorithms_demo.rb
def bfs(graph, start)
  order = []
  seen = {}
  queue = [start]
  until queue.empty?
    node = queue.shift
    next if seen[node]

    seen[node] = true
    order << node
    queue.concat(graph.fetch(node, []))
  end
  order
end

scores = [["Lin", 91], ["Ada", 95], ["Kai", 91]]
ranked = scores.sort_by { |name, score| [-score, name] }
numbers = [2, 5, 8, 13, 21]
graph = { "A" => %w[B C], "B" => ["D"], "C" => ["D"] }

puts "sorted=#{ranked.map { |name, score| "#{name}:#{score}" }.join(',')}"
puts "binary-search-13=#{numbers.bsearch_index { |value| value >= 13 }}"
puts "bfs=#{bfs(graph, 'A').join(',')}"
